src/core/thread/network_data_leader.cpp

```diff
@@ -577,20 +575,200 @@ void Leader::SendServerDataResponse(const Coap::Header &aRequestHeader, const Ip
+    NetworkDataTlv *subEnd;
+    PrefixTlv *prefix;
+    BorderRouterTlv *borderRouter;
+    HasRouteTlv *hasRoute;
+    BorderRouterEntry *borderRouterEntry;
+    HasRouteEntry *hasRouteEntry;
+    *aIn = false;
+    *aStable = false;
+
+    while (1)
+    {
+        if (cur >= end)
+        {
+            break;
+        }
+
+        if (cur->GetType() == NetworkDataTlv::kTypePrefix)
+        {
+            prefix = reinterpret_cast<PrefixTlv *>(cur);
+            subCur = reinterpret_cast<NetworkDataTlv *>(prefix->GetSubTlvs());
+            subEnd = reinterpret_cast<NetworkDataTlv *>(prefix->GetSubTlvs() + prefix->GetSubTlvsLength());
+
+            while (1)
+            {
+                if (subCur >= subEnd)
+                {
+                    break;
+                }
+
+                switch (subCur->GetType())
+                {
+                case NetworkDataTlv::kTypeBorderRouter:
+                    borderRouter = FindBorderRouter(*prefix);
+                    for (int i = 0; i < borderRouter->GetNumEntries(); i++)
+                    {
+                        borderRouterEntry = borderRouter->GetEntry(i);
+                        if (borderRouterEntry->GetRloc() == aRloc16)
+                        {
+                            *aIn = true;
+                            if (borderRouter->IsStable())
+                            {
+                                *aStable = true;
+                            }
+                        }
+                    }
+                    break;
+                case NetworkDataTlv::kTypeHasRoute:
+                    hasRoute = FindHasRoute(*prefix);
+                    for (int i = 0; i < hasRoute->GetNumEntries(); i++)
+                    {
+                        hasRouteEntry = hasRoute->GetEntry(i);
+                        if (hasRouteEntry->GetRloc() == aRloc16)
+                        {
+                            *aIn = true;
+                            if (hasRoute->IsStable())
+                            {
+                                *aStable = true;
+                            }
+                        }
+                    }
+                    break;
+                default:
+                    break;
+                }
+
+                if ((*aIn) && (*aStable))
+                {
+                    ExitNow();
+                }
+
+                subCur = subCur->GetNext();
+            }
+        }
+        cur = cur->GetNext();
+    }
+
+exit:
+    return;
+}
+
+bool Leader::IsStableUpdated(uint16_t aRloc16, uint8_t *aTlvs, uint8_t aTlvsLength, uint8_t *aTlvsBase, uint8_t aTlvsBaseLength)
+{
+    bool rval = false;
+    NetworkDataTlv *cur = reinterpret_cast<NetworkDataTlv *>(aTlvs);
+    NetworkDataTlv *end = reinterpret_cast<NetworkDataTlv *>(aTlvs + aTlvsLength);
+    PrefixTlv *prefix;
+    PrefixTlv *prefixBase;
+    BorderRouterTlv *borderRouter;
+    HasRouteTlv *hasRoute;
+
+    while (1)
+    {
+        if (cur >= end)
```

Comment:
Why not put `cur < end` in the `while` statement above and remove this `if` statement?